pkg/pillar/cas/containerd.go

```diff
@@ -167,6 +159,23 @@ func (c *containerdCAS) IngestBlob(blobs ...*types.BlobStatus) ([]*types.BlobSta
 			sha = fmt.Sprintf("%s:%s", digest.SHA256, strings.ToLower(blob.Sha256))
 		)
 
+		log.Infof("IngestBlob(%s): processing blob %+v", blob.Sha256, blob)
+		// Process the blob only if its not in a loaded status already
+		if blob.State == types.LOADED {
+			log.Infof("IngestBlob(%s): Not loading blob as it is already marked as loaded", blob.Sha256)
+			loadedBlobs = append(loadedBlobs, blob)
+			continue
+		}
+
+		// Process the blob only if it isn't already loaded. It might be loaded but not
```

Comment:
Don't you need to explicitly check for LOADING here? containerd might not have started working on it yet, but volumemgr might have marked it as LOADING and told containerd to get working on it.

Reply:
Thus explicitly asking containerd seems to miss half the race conditions.

Reply:
And if it is LOADING it can't be added to loadedBlobs since we need to wait for it to become LOADED

Reply:
I agree that this only covers part of the race conditions. The question is, how would you cover the others? Specifically, if it is in state LOADING:

* how do we check that state? do we look at blob status from within containerd package real-time?
* more importantly, if we see that it is LOADING, what do we do here? It obviously is LOADING from some other parallel process (race condition), so what do we do here? wait? create a new loop? load the other blobs and return success, but have the main loop not go back until everything is loaded?

Reply:
blobStatus.State == LOADING works as  check.
We don't care if containerd is in vacation for a week and will not start processing until back; we care that we have asked it to do the work and presumably blobStatus.State == LOADING captures that.

When the state is LOADING the work needed is no different than when we find that the state is < LOADING and we make it be LOADING. I haven't checked how the code handles that case. But the end result is that the blob doesn't get into loadedBlobs hence we end up concluding we are not done yet.

Reply:
So what are you suggesting? I think the following:

1. volumemgr submits a job: "load these blobs and create an image"
1. worker receives it and tries to load all
   * If any "true" errors, it returns an error
   * If any blobs already loaded (`BlobStatus.State == LOADED`), load the rest and return ok (we already do this)
   * If any blobs are loading (`BlobStatus.State == LOADING`), assume something else is loading it, load the rest, create the image, return an "ok but not done yet"
1. If the returned result was "ok but not done yet", loop around again until it returns "successfully loaded all blobs or found them loaded"

Is that it? The job could have _three_ return states:

* error
* success
* incomplete, try me again later

and the call would be idempotent.

Is that about it?